### PR TITLE
Better handling of edge cases in DateInput data

### LINKED ISSUE #340: Regression in DateInput
Hi guys!

I got a regression in updating from 1.1.0 to 1.2.0 (latest).

The bug comes from [this PR](https://github.com/palantir/blueprint/pull/297).

The component crashes during its instantiation because of [this change](https://github.com/palantir/blueprint/pull/297/files#diff-2899fd9fc14d599c6e4062f791d99120L135).

Some screenshots:

screen shot 2016-12-09 at 00 47 57

screen shot 2016-12-09 at 00 47 24

It worked before to pass a `string` to the component because of the `moment(date)`.

Now it doesn't work anymore because of the `moment([date.getFullYear(), ...])`.

To be honest, your doc says that `value` should be a `Date` and not a `string`, so...

We could indicate in the changelog the potential breaking change and a possible solution, e.g.

> replacing `value={value}` by `value={new Date(value)}`

Or maybe a workaround in the code to keep the previous behaviour?

### PR DESCRIPTION
**Pull request by adidahiya**
#### Checklist

- [x] Include tests
- [N/A] Update documentation

#### Changes proposed in this pull request:

Check for nulls and strings in `fromMomentToDate` and `fromDateToMoment`

### CODE REVIEW

**Comment by adidahiya** [comment]
Added some tests, but they're not the most robust... leaving it up to you if you want to improve them; I just wanted to get a quick fix in

File: packages/datetime/test/dateInputTests.tsx [hunk 1]
```diff
@@ -14,6 +14,14 @@ import { padWithZeroes } from "../src/common/utils";
 import { Classes, DateInput } from "../src/index";
 
 describe("<DateInput>", () => {
+    it("handles null inputs without crashing", () => {
+        mount(<DateInput value={null} />);
```
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]:
```js
assert.throws(() => mount(...))
```

https://github.com/palantir/blueprint/blob/master/packages/core/test/tabs/tabsTests.tsx#L46

Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: or actually `doesNotThrow`

Comment (adidahiya) [inline packages/datetime/test/dateInputTests.tsx]: That should be implicit in the tests; we shouldn't allow uncaught exceptions. If that's not the case we should fix that soon...

Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: ok fair, i guess `assert.doesNotThrow` lets you validate the type of exception but we care about _any_ exception. this is reasonable.

File: packages/datetime/test/dateInputTests.tsx [hunk 2]
```diff
@@ -14,6 +14,14 @@ import { padWithZeroes } from "../src/common/utils";
...
+    });
+
+    it("handles string inputs without crashing", () => {
+        mount(<DateInput value="1988-08-07 11:01:12" />);
```
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: @adidahiya this is not valid usage of the component and throws a compiler error in tests. should we add `string` to `value` types or cast this `as any`?

File: packages/datetime/test/dateInputTests.tsx (hunk 1 above)
Comment (cmslewis) [inline packages/datetime/test/dateInputTests.tsx]: I find `assert.doesNotThrow` a little clearer than having nothing at all, but I don't feel strongly.

File: packages/datetime/test/dateInputTests.tsx (hunk 2 above)
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: from latest circle build: 
image

File: packages/datetime/test/dateInputTests.tsx (hunk 1 above)
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: tested locally, this is reasonable.

File: packages/datetime/src/dateInput.tsx [hunk 3]
```diff
@@ -327,34 +327,38 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
...
-            momentDate.year(),
-            momentDate.month(),
-            momentDate.date(),
-            momentDate.hours(),
-            momentDate.minutes(),
-            momentDate.seconds(),
-            momentDate.milliseconds(),
-        );
+        if (momentDate == null) {
+            return undefined;
```
Comment (cmslewis) [inline packages/datetime/src/dateInput.tsx]: nit: Do we care about returning `undefined` for an `undefined` param and `null` for a `null` param? That's the pattern I tend to prefer. Not a big deal to me though.

File: packages/datetime/test/dateInputTests.tsx (hunk 1 above)
Comment (giladgray) [inline packages/datetime/test/dateInputTests.tsx]: i agree on the clarity @cmslewis. it's not obvious that we didn't simply forget to finish the test case here.

**Pull request merged**